File: guests/java/dev/kaya/guests/Ownundo.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

import java.util.ArrayList;
import java.util.List;

/**
 * The app-owned undo scene from the JVM — guests/rust/ownundo.rs,
 * tools/scenes/ownundo.steps (docs/rich-text-plan.md R6, §14).
 */
public final class Ownundo {
    /** The app's own history: the document before each user edit, and the
     * documents an undo took away. A named holder, not static fields —
     * this scene's history has no reason to outlive its one app(). */
    /** The app's OWN undo history, plus the two textareas its menu
     * items were declared before: mutable app state and a forward
     * reference, not a container body smuggling its result out. */
    private static final class State {
        KayaApp.Widget nativeArea, owned;
        final List<KayaApp.Document> undo = new ArrayList<>();
        final List<KayaApp.Document> redo = new ArrayList<>();
        KayaApp.Document current = new KayaApp.Document("");
    }

    public static void app() {
        KayaApp app = new KayaApp();

        State s = new State();

        app.build(tx -> {
            KayaApp.WindowRef win = tx.window(0).title("ownundo");
            KayaApp.MenuItem edit = win.menu("Edit");

            KayaApp.Signal<String> status = tx.signal("undo 0 redo 0");

            edit.item("Undo").role(KayaApp.ROLE_UNDO).onActivate(t -> {
                if (s.undo.isEmpty()) {
                    return;
                }
                KayaApp.Document before = s.undo.remove(s.undo.size() - 1);
                s.redo.add(s.current);
                s.current = before;
                t.setDocument(s.owned, before);
                publish(t, status, s);
            });
            edit.item("Redo").role(KayaApp.ROLE_REDO).onActivate(t -> {
                if (s.redo.isEmpty()) {
                    return;
                }
                KayaApp.Document after = s.redo.remove(s.redo.size() - 1);
                s.undo.add(s.current);
                s.current = after;
                t.setDocument(s.owned, after);
                publish(t, status, s);
            });

            tx.mount(tx.column(col -> {
                tx.label(status).a11yId("status"); // label#0
                s.nativeArea = tx.textarea().rich().a11yId("native").a11yLabel("Native");
                s.owned = tx.textarea().rich().ownUndo()
                        .a11yId("owned").a11yLabel("Owned");
                app.onEdit(s.owned, (t, e) -> {
                    s.undo.add(s.current);
                    s.current = app.document(s.owned);
                    s.redo.clear();
                    publish(t, status, s);
                });
                tx.row(row -> {
                    tx.button("focus native", t -> t.focus(s.nativeArea)); // button#0
                    tx.button("focus owned", t -> t.focus(s.owned));  // button#1
                });
            }));
        });

        app.dispatchLoop();
    }

    private static void publish(KayaApp.Tx t, KayaApp.Signal<String> status, State s) {
        t.write(status, "undo " + s.undo.size() + " redo " + s.redo.size());
        t.canUndo(s.owned, !s.undo.isEmpty());
        t.canRedo(s.owned, !s.redo.isEmpty());
    }

    private Ownundo() {}
}

File: guests/go/ownundo/ownundo.go
// The app-owned undo scene (tools/scenes/ownundo.steps;
// docs/rich-text-plan.md R6, §14): two rich textareas, one on the
// platform's own undo tier and one whose app keeps its own history of
// Documents. Edit>Undo reaches the app through the role item's own
// activation while the owned textarea is focused, and the platform's tier
// while the other one is.
package ownundo

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

func App() *kaya.App {
	app := kaya.NewApp()

	// THE APP'S OWN HISTORY: the document before each user edit, and the
	// documents an undo took away. The binding's mirror is the document
	// AFTER the edit it just delivered.
	var undo, redo []kaya.Document
	current := kaya.NewDocument("")

	var status kaya.Signal[string]
	var native, owned kaya.Widget

	publish := func(tx *kaya.Tx) {
		tx.Write(status, fmt.Sprintf("undo %d redo %d", len(undo), len(redo)))
		tx.CanUndo(owned, len(undo) > 0)
		tx.CanRedo(owned, len(redo) > 0)
	}

	app.Build(func(tx *kaya.Tx) {
		win := tx.Window(0).Title("ownundo")

		edit := win.Menu("Edit")
		edit.Item("Undo").Role(kaya.RoleUndo).OnActivate(func(tx *kaya.Tx) {
			if len(undo) == 0 {
				return
			}
			before := undo[len(undo)-1]
			undo = undo[:len(undo)-1]
			redo = append(redo, current)
			current = before
			tx.SetDocument(owned, before)
			publish(tx)
		})
		edit.Item("Redo").Role(kaya.RoleRedo).OnActivate(func(tx *kaya.Tx) {
			if len(redo) == 0 {
				return
			}
			after := redo[len(redo)-1]
			redo = redo[:len(redo)-1]
			undo = append(undo, current)
			current = after
			tx.SetDocument(owned, after)
			publish(tx)
		})

		status = tx.Signal("undo 0 redo 0")

		tx.Mount(tx.Column(func() {
			tx.Label(status).A11yID("status") // label#0
			native = tx.Textarea(nil).Rich().
				A11yID("native").A11yLabel("Native") // textarea#0
			owned = tx.Textarea(nil).Rich().OwnUndo().
				A11yID("owned").A11yLabel("Owned") // textarea#1
			owned.OnEdit(func(tx *kaya.Tx, _ kaya.Edit) {
				undo = append(undo, current)
				current = app.Document(owned)
				redo = redo[:0]
				publish(tx)
			})

			tx.Row(func() {
				tx.Button("focus native", func(tx *kaya.Tx) { // button#0
					tx.Focus(native)
				})
				tx.Button("focus owned", func(tx *kaya.Tx) { // button#1
					tx.Focus(owned)
				})
			})
		}))
	})

	return app
}
